src/rate/index.tsx

```diff
@@ -0,0 +1,336 @@
+import { DNode, RenderResult } from '@dojo/framework/core/interfaces';
+import { create, tsx } from '@dojo/framework/core/vdom';
+import focus from '@dojo/framework/core/middleware/focus';
+import i18n from '@dojo/framework/core/middleware/i18n';
+import { createICacheMiddleware } from '@dojo/framework/core/middleware/icache';
+import Icon from '../icon';
+import Radio from '../radio';
+import RadioGroup, { RadioGroupChildren } from '../radio-group';
+import theme from '../middleware/theme';
+import * as baseCss from '../common/styles/base.m.css';
+import * as iconCss from '../theme/default/icon.m.css';
+import * as radioCss from '../theme/default/radio.m.css';
+import * as radioGroupCss from '../theme/default/radio-group.m.css';
+import * as css from '../theme/default/rate.m.css';
+import bundle from './nls/Rate';
+
+export interface MixedNumber {
+	value: number;
+	quotient: number;
+	numerator: number;
+	denominator: number;
+}
+
+export interface RateProperties {
+	/** The name attribute for this rating group */
+	name: string;
+	/** The label to be displayed in the legend */
+	label?: string;
+	/** The initial rating value */
+	initialValue?: number;
+	/** Callback fired when the rating value changes */
+	onValue?: (value?: number) => void;
+	/** The number of times each star is divided */
+	steps?: number;
+	/** The highest rating for this rating group */
+	max?: number;
+	/** Allow undefined to be chosen with repeat click or with arrows */
+	allowClear?: boolean;
+	/** Prevent interaction with a visual indication */
+	disabled?: boolean;
+	/** Prevent interaction */
+	readOnly?: boolean;
+}
+
+export interface RateChildren {
+	(fill: boolean, integer: number, selected?: number, over?: number): RenderResult;
+}
+
+interface RateState {
+	uuid: string;
+	hover: string;
+	hovering: boolean;
+	selectedInteger?: number;
+	selectedStep?: number;
+	focused: boolean;
+}
+
+const icache = createICacheMiddleware<RateState>();
+
+function mixedNumber(integer: number, step: number, steps: number): MixedNumber {
+	if (step === steps) {
+		step = 0;
+	} else {
+		integer -= 1;
+	}
+	return {
+		value: integer + step / steps,
+		quotient: integer,
+		numerator: step,
+		denominator: steps
+	};
+}
+
+type RadioChecked = (checked?: boolean) => boolean | undefined;
+
+const factory = create({ focus, i18n, icache, theme })
+	.properties<RateProperties>()
+	.children<RateChildren | undefined>();
+
+export const Rate = factory(function Rate({
+	id,
+	children,
+	properties,
+	middleware: { focus, i18n, icache, theme }
+}) {
+	const { format } = i18n.localize(bundle);
+	const themedCss = theme.classes(css);
+	const {
+		name,
+		label,
+		initialValue: initialNumber = undefined,
```

Comment:
is there a reason this is renamed to `initialNumber`? I don't think there's any need to default this to undefined.

Reply:
I treat it internally as the radio group's initial value.